src/S3/S3Transfer/MultipartUploader.php

```diff
@@ -0,0 +1,542 @@
+            $partNo++;
+            $read = $this->body->read($partSize);
+            // To make sure we do not create an empty part when
+            // we already reached the end of file.
+            if (empty($read) && $this->body->eof()) {
+                break;
+            }
+
+            $partBody  = Utils::streamFor(
+                $read
+            );
+            $uploadPartCommandArgs = [
+                ...$this->createMultipartArgs,
+                'UploadId' => $this->uploadId,
+                'PartNumber' => $partNo,
+                'ContentLength' => $partBody->getSize(),
+            ];
+            // To get `requestArgs` when notifying the bytesTransfer listeners.
+            $uploadPartCommandArgs['requestArgs'] = [...$uploadPartCommandArgs];
+            // Attach body
+            $uploadPartCommandArgs['Body'] = $this->decorateWithHashes(
+                $partBody,
+                $uploadPartCommandArgs
+            );
+            $command = $this->s3Client->getCommand('UploadPart', $uploadPartCommandArgs);
+            $commands[] = $command;
+            $this->calculatedObjectSize += $partBody->getSize();
+            if ($partNo > self::PART_MAX_NUM) {
+                return Create::rejectionFor(
+                    "The max number of parts has been exceeded. " .
+                    "Max = " . self::PART_MAX_NUM
+                );
+            }
+        }
+
+        return (new CommandPool(
+            $this->s3Client,
+            $commands,
+            [
+                'concurrency' => $this->config['concurrency'],
+                'fulfilled'   => function (ResultInterface $result, $index)
+                    use ($commands) {
+                        $command = $commands[$index];
+                        $this->collectPart(
+                            $result,
+                            $command
+                        );
+                        // Part Upload Completed Event
+                        $this->partUploadCompleted(
+                            $command['ContentLength'],
+                            $command['requestArgs']
+                        );
+                },
+                'rejected'     => function (Throwable $e) {
+                    $this->partUploadFailed($e);
+
+                    throw $e;
+                }
+            ]
+        ))->promise();
+    }
+
+    /**
+     * @return PromiseInterface
+     */
+    private function completeMultipartUpload(): PromiseInterface
+    {
+        $this->sortParts();
+        $completeMultipartUploadArgs = [
+            ...$this->createMultipartArgs,
+            'UploadId' => $this->uploadId,
+            'MpuObjectSize' => $this->calculatedObjectSize,
+            'MultipartUpload' => [
+                'Parts' => $this->parts,
+            ]
+        ];
+        if ($this->containsChecksum($this->createMultipartArgs)) {
+            $completeMultipartUploadArgs['ChecksumType'] = 'FULL_OBJECT';
```

Comment:
We should have some assertions on the resulting `x-amz-checksum-type` header ensuring this value is provided.  Also, I think we need to set this in the `CreateMultipartUpload` call as well.